File: 757-pyramid-transition-matrix/pyramid-transition-matrix.java
import java.util.*;

class Solution {

    // Class-level variables
    private Map<String, List<Character>> map;
    private Map<String, Boolean> memo;

    public boolean pyramidTransition(String bottom, List<String> allowed) {

        map = new HashMap<>();
        memo = new HashMap<>();

        // Build transition map
        for (String s : allowed) {
            String key = s.substring(0, 2);
            map.computeIfAbsent(key, k -> new ArrayList<>()).add(s.charAt(2));
        }

        return backtrack(bottom.length(), 0, 1, bottom, new StringBuilder());
    }

    private boolean backtrack(
            int stage,
            int i,
            int j,
            String bottom,
            StringBuilder nxt) {

        // Base case
        if (stage == 1) {
            return true;
        }

        // ✅ Memoization on FULL ROW only
        if (i == 0 && j == 1 && memo.containsKey(bottom)) {
            return memo.get(bottom);
        }

        // ✅ Fail-fast pruning on FULL ROW
        if (i == 0 && j == 1) {
            for (int x = 0; x < bottom.length() - 1; x++) {
                if (!map.containsKey(bottom.substring(x, x + 2))) {
                    memo.put(bottom, false);
                    return false;
                }
            }
        }

        String pair = "" + bottom.charAt(i) + bottom.charAt(j);

        if (map.containsKey(pair)) {
            for (char ch : map.get(pair)) {
                nxt.append(ch);

                // Finished building next row
                if (j == stage - 1) {
                    if (backtrack(stage - 1, 0, 1,
                                  nxt.toString(), new StringBuilder())) {
                        memo.put(bottom, true);
                        return true;
                    }
                } else {
                    if (backtrack(stage, i + 1, j + 1,
                                  bottom, nxt)) {
                        return true;
                    }
                }

                nxt.deleteCharAt(nxt.length() - 1);
            }
        }

        // Cache negative result for full row
        if (i == 0 && j == 1) {
            memo.put(bottom, false);
        }

        return false;
    }
}

File: 757-pyramid-transition-matrix/pyramid-transition-matrix.py
from typing import List
from collections import defaultdict

class Solution:
    def pyramidTransition(self, bottom: str, allowed: List[str]) -> bool:
        # Build transition map
        trans = defaultdict(list)
        for s in allowed:
            trans[s[:2]].append(s[2])

        memo = {}

        def dfs(curr: str) -> bool:
            # Base case
            if len(curr) == 1:
                return True

            # Memo check
            if curr in memo:
                return memo[curr]

            # FAIL-FAST PRUNING
            for i in range(len(curr) - 1):
                if curr[i:i+2] not in trans:
                    memo[curr] = False
                    return False

            def build(idx: int, nxt: List[str]) -> bool:
                if idx == len(curr) - 1:
                    return dfs("".join(nxt))

                pair = curr[idx:idx+2]
                for c in trans[pair]:
                    nxt.append(c)
                    if build(idx + 1, nxt):
                        return True
                    nxt.pop()
                return False

            memo[curr] = build(0, [])
            return memo[curr]

        return dfs(bottom)
